### PR TITLE
Add an API to set the number of threads in C10 thread pool

### PR DESCRIPTION
**Pull request by jamesr66a**
Tested locally on machine translation service

### CODE REVIEW

File: aten/src/ATen/core/thread_pool.cpp [hunk 1]
```diff
@@ -1,16 +1,16 @@
-#include <ATen/core/thread_pool.h>
 #include <ATen/core/ivalue.h>
+#include <ATen/core/thread_pool.h>
 
 namespace c10 {
 
-ThreadPool::ThreadPool(std::size_t pool_size, int numa_node_id)
-    : threads_(pool_size),
+ThreadPool::ThreadPool(c10::optional<std::size_t> pool_size, int numa_node_id)
+    : threads_(pool_size.has_value() ? pool_size.value() : 1),
```
Comment (dzhulgakov) [inline aten/src/ATen/core/thread_pool.cpp]: .value_or(1)?

File: aten/src/ATen/core/thread_pool.cpp [hunk 2]
```diff
@@ -117,8 +117,18 @@ void ThreadPool::main_loop(std::size_t index) {
   } // while running_
 }
 
+c10::optional<size_t> num_threads(c10::optional<size_t> val) {
```
Comment (dzhulgakov) [inline aten/src/ATen/core/thread_pool.cpp]: it's a bit strange API to both get and set, but... ok

File: aten/src/ATen/core/thread_pool.cpp [hunk 3]
```diff
@@ -117,8 +117,18 @@ void ThreadPool::main_loop(std::size_t index) {
...
+  static c10::optional<size_t> n_threads = val;
+  if (val.has_value() && val != n_threads) {
+    throw std::runtime_error(
```
Comment (dzhulgakov) [inline aten/src/ATen/core/thread_pool.cpp]: log values?

File: aten/src/ATen/core/thread_pool.cpp [hunk 4]
```diff
@@ -117,8 +117,18 @@ void ThreadPool::main_loop(std::size_t index) {
   } // while running_
 }
 
+c10::optional<size_t> num_threads(c10::optional<size_t> val) {
+  static c10::optional<size_t> n_threads = val;
```
Comment (dzhulgakov) [inline aten/src/ATen/core/thread_pool.cpp]: so what happens if someone calls this function first to get num threads? it'd be none forever? imho, it's better to default to 1 here there in that case and remove the conditional in ThreadPool.

**Review by zdevito — APPROVED** [review]
Looks fine. I have one suggestion to ensure setting the number of threads can only happen before the first call to get the engine.

**Pull request closed without merging**